### PR TITLE
[MGCB] changes to output for Visual Studio

### PR DESCRIPTION
**Pull request by nkast**
Formatting the Output of a Custom Build Step or Build Event
https://msdn.microsoft.com/en-us/library/yxkt8b26.aspx

_Before_
mgerrorlist0

```
3>  P:/TestContentBuildLogger/MG/Content/test.fbx
3>      LogMessage
3>  LogWarning0
3>  P:/TestContentBuildLogger/MG/Content/test.fbx: LogWarning
3>  P:/TestContentBuildLogger/MG/Content/test.fbx(64,4): LogWarningWithLineNum
3>  P:/TestContentBuildLogger/MG/Content/test.fbx(64,4): LogWarningWithLink
3>  P:/TestContentBuildLogger/MG/Content/test.fbx(64,4): LogWarningWithNewLine
3>  WinNewLine
3>      LogImportantMessage
3>      LogImportantMessageWithNewLine
3>  WinNewLine
3>  Console.WriteLine
3>  Error.WriteLine
3>  P:/TestContentBuildLogger/MG/Content/test.fbx(42,8): TestInvalidContentException
3>  Newline1
3>  Newline2
3>  Return1
3>  Return2
3>  NewlineAgain
3>  WinNewLine
3>  P:/TestContentBuildLogger/MG/Content/test1.fbx
3>P:/TestContentBuildLogger/MG/Content/test1.fbx : error : Processor 'GpuSkinnedModelProcessor' had unexpected failure!
3>  System.NullReferenceException: test1_NullReferenceException
3>     at tainicom.ProtonType.Content.Pipeline.GpuSkinnedModelProcessor.Process(NodeContent input, ContentProcessorContext context) in p:\tainicom\Pinball_League\EspressoRub\trunk\Skinning\SkinnedModelImporter\GpuSkinnedModelProcessor.cs:line 43
3>     at Microsoft.Xna.Framework.Content.Pipeline.ContentProcessor`2.Microsoft.Xna.Framework.Content.Pipeline.IContentProcessor.Process(Object input, ContentProcessorContext context) in P:\MonoGame\MonoGame.Framework.Content.Pipeline\ContentProcessor.cs:line 60
3>     at MonoGame.Framework.Content.Pipeline.Builder.PipelineManager.ProcessContent(PipelineBuildEvent pipelineEvent) in P:\MonoGame\MonoGame.Framework.Content.Pipeline\Builder\PipelineManager.cs:line 696
3>C:\Program Files (x86)\MSBuild\MonoGame\v3.0\MonoGame.Content.Builder.targets(91,5): error MSB3073: The command ""C:\Program Files (x86)\MSBuild\MonoGame\v3.0\Tools\MGCB.exe" /@:"P:\TestContentBuildLogger\MG\Content\Content.mgcb" /platform:Windows /outputDir:"P:\TestContentBuildLogger\MG\Content\bin\Windows" /intermediateDir:"P:\TestContentBuildLogger\MG\Content\obj\Windows" /quiet" exited with code 2.
```

_after_
mgerrorlist1

```
3>P:/TestContentBuildLogger/MG/Content/test.fbx
3>      LogMessage
3>P:/TestContentBuildLogger/MG/Content/test.fbx : warning : LogWarning0
3>P:/TestContentBuildLogger/MG/Content/test.fbx : warning : LogWarning
3>P:/TestContentBuildLogger/MG/Content/test.fbx(64,4): warning : LogWarningWithLineNum
3>P:/TestContentBuildLogger/MG/Content/test.fbx(64,4): warning : LogWarningWithLink
3>P:/TestContentBuildLogger/MG/Content/test.fbx(64,4): warning : LogWarningWithNewLine
3>  WinNewLine
3>      LogImportantMessage
3>      LogImportantMessageWithNewLine
3>  WinNewLine
3>  Console.WriteLine
3>  Error.WriteLine
3>P:/TestContentBuildLogger/MG/Content/test.fbx : error : TestInvalidContentException
3>  Newline1
3>  Newline2
3>  Return1
3>  Return2
3>  NewlineAgain
3>  WinNewLine
3>  P:/TestContentBuildLogger/MG/Content/test1.fbx
3>P:/TestContentBuildLogger/MG/Content/test1.fbx : error : Processor 'GpuSkinnedModelProcessor' had unexpected failure!
3>  System.NullReferenceException: test1_NullReferenceException
3>     at tainicom.ProtonType.Content.Pipeline.GpuSkinnedModelProcessor.Process(NodeContent input, ContentProcessorContext context) in p:\tainicom\Pinball_League\EspressoRub\trunk\Skinning\SkinnedModelImporter\GpuSkinnedModelProcessor.cs:line 39
3>     at Microsoft.Xna.Framework.Content.Pipeline.ContentProcessor`2.Microsoft.Xna.Framework.Content.Pipeline.IContentProcessor.Process(Object input, ContentProcessorContext context) in P:\MonoGame\MonoGame.Framework.Content.Pipeline\ContentProcessor.cs:line 60
3>     at MonoGame.Framework.Content.Pipeline.Builder.PipelineManager.ProcessContent(PipelineBuildEvent pipelineEvent) in P:\MonoGame\MonoGame.Framework.Content.Pipeline\Builder\PipelineManager.cs:line 696
3>C:\Program Files (x86)\MSBuild\MonoGame\v3.0\MonoGame.Content.Builder.targets(91,5): error MSB3073: The command ""C:\Program Files (x86)\MSBuild\MonoGame\v3.0\Tools\MGCB.exe" /@:"P:\TestContentBuildLogger\MG\Content\Content.mgcb" /platform:Windows /outputDir:"P:\TestContentBuildLogger\MG\Content\bin\Windows" /intermediateDir:"P:\TestContentBuildLogger\MG\Content\obj\Windows" /quiet" exited with code 2.
```

_XNA output_
xnaerrorlist

```
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : warning : The skeleton root, "PIVOT", has been pulled out of the scene hierarchy and made into a child of the root node. One of the skeleton root's old parents, "PersonSupervisor", has an animation named "Base Stack" which will not be reflected on the skeleton root.
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : warning : Fragment identifier "PersonSupervisor".
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : warning : LogWarning0
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : warning : LogWarning
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx(64,4): warning : LogWarningWithLineNum
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx(64,4): warning : LogWarningWithLink
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx(64,4): warning : LogWarningWithNewLine
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx(64,4): warning : WinNewLine
  LogImportantMessage
  LogImportantMessageWithNewLine
  WinNewLine
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : error : TestInvalidContentException
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : error : Newline1
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : error : Newline2
Return1
Return2
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : error : NewlineAgain
P:\TestContentBuildLogger\XNA\ContentTest\ContentTestContent\test.fbx : error : WinNewLine
```

_test code_

```
            if(input.Identity.SourceFilename.Contains("test1"))
                throw new System.NullReferenceException("test1_NullReferenceException");
            context.Logger.LogMessage("LogMessage");
            context.Logger.LogWarning(null, null, "LogWarning0");
            context.Logger.LogWarning(null, input.Identity, "LogWarning");
            input.Identity.FragmentIdentifier = "64,4";
            context.Logger.LogWarning(null, input.Identity, "LogWarningWithLineNum");
            context.Logger.LogWarning("http://www.monogame.net/", input.Identity, "LogWarningWithLink");
            context.Logger.LogWarning("http://www.monogame.net/", input.Identity, "LogWarningWithNewLine\r\nWinNewLine");
            context.Logger.LogImportantMessage("LogImportantMessage");
            context.Logger.LogImportantMessage("LogImportantMessageWithNewLine\r\nWinNewLine");
            System.Console.WriteLine("Console.WriteLine");            
            System.Console.Error.WriteLine("Error.WriteLine");

            input.Identity.FragmentIdentifier = "42,8";
            input.Identity.SourceTool = "customTool v 0.45a";
            if (input.Identity.SourceFilename.Contains("test"))
                throw new InvalidContentException("TestInvalidContentException\nNewline1\nNewline2\rReturn1\rReturn2\nNewlineAgain\r\nWinNewLine", input.Identity);
```

### CODE REVIEW

**Comment by nkast** [comment]
I just found that it doesn't work very well with EffectProcessor, Pipeline filterView fails to parse the line, column numbers and VS leaves the errorCode in the description.

Now `LogWarning()` adds the ": warning : " to the output and EffecrProcessor adds the message "warning X3206: msg" add the end. We have to pass that ErrorCode somehow... 

`P:\\MG Tests\\Test\\Content\\File.fx(31,12): warning : warning X3206: implicit truncation of vector type`


https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Processors/EffectProcessor.cs#L211

**Comment by nkast** [comment]
For errors we can add a new `ErrorCode` property to `InvalidContentException`, but for warning we are left with some bad options/hacks.

-Add a new abstract method to `ContentBuildLogger`.
-Pass the errorCode in the message in the form of "X3206:implicit truncation of vector type" and then Console parse the message with a strict RE like "[A-Z][09]*:"

**Comment by nkast** [comment]
Ok, MGFX messages fixed.

**Comment by KonajuGames** [comment]
@nkast Could this be updated to latest?

**Comment by nkast** [comment]
@KonajuGames done.

**Comment by nkast** [comment]
NOTE: need to test/update this with PR #5849

**Comment by nkast** [comment]
Recent changes to the pipeline drift this PR apart from develop. Some warning-errors don't show up in the VS Error panel or are show incorrectly.

File: Tools/MGCB/BuildContent.cs [hunk 1]
```diff
@@ -336,22 +337,33 @@ public void Build(out int successCount, out int errorCount)
...
+                    message += ": error ";
+
+                    // extract errorCode from MGFX message
+                    var match = Regex.Match(ex.Message, @"([A-Z]*[0-9]*):(.*)");
+                    if (match.Success || match.Groups.Count == 2)
+                        message += match.Groups[1].Value + " : " + match.Groups[2].Value;
+                    else
+                        message += ": " + ex.Message;
+
                     Console.WriteLine(message);
```
Comment (SeriousMartin) [inline Tools/MGCB/BuildContent.cs]: Shouldn't this Console.Error.WriteLine as-well?

Comment (nkast) [reply Tools/MGCB/BuildContent.cs]: This sound right.
I didn't change more code than I had to in order to make it work.
Yet, looking back at code I wrote a year ago I can't comment on anything! I am looking at it fresh.

**Pull request closed without merging**